nogo panic

I'm trying (again) to enable nogo but it panics:
```console
 bazel test //... -- -//vendor/...
INFO: Invocation ID: b3f0a5c8-76f5-4f8e-a138-1d1ad566a993
DEBUG: /private/var/tmp/_bazel_ash2k/18feaf204f522c9679e75c03985bfc8f/external/bazel_skylib/lib.bzl:30:1: WARNING: lib.bzl is deprecated and will go away in the future, please directly load the bzl file(s) of the module(s) needed as it is more efficient.
DEBUG: Rule 'org_golang_x_tools' modified arguments {"sha256": "11629171a39a1cb4d426760005be6f7cb9b4182e4cb2756b7f1c5c2b6ae869fe"}
INFO: Analysed 44 targets (1332 packages loaded, 2972 targets configured).
INFO: Found 35 targets and 9 test targets...
ERROR: /Users/ash2k/go/src/github.com/atlassian/smith/vendor/k8s.io/klog/BUILD.bazel:3:1: GoCompile vendor/k8s.io/klog/darwin_amd64_stripped/go_default_library%/github.com/atlassian/smith/vendor/k8s.io/klog.a failed (Exit 1) compile failed: error executing command bazel-out/host/bin/external/io_bazel_rules_go/go/tools/builders/darwin_amd64_stripped/compile -sdk external/go_sdk -installsuffix darwin_amd64 -src vendor/k8s.io/klog/klog.go -src ... (remaining 14 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox
GoCompile: nogo: internal error: gob encoding of analysis fact printWrapper failed: gob: type not registered for interface: printf.isWrapper; please report a bug against fact *printf.isWrapper in package "golang.org/x/tools/go/analysis/passes/printf"
panic: internal error: gob encoding of analysis fact printWrapper failed: gob: type not registered for interface: printf.isWrapper; please report a bug against fact *printf.isWrapper in package "golang.org/x/tools/go/analysis/passes/printf"

goroutine 1 [running]:
log.Panicf(0x12a6b1d, 0x6e, 0xc00018bc78, 0x4, 0x4)
        GOROOT/src/log/log.go:333 +0xda
golang.org/x/tools/go/analysis/internal/facts.(*Set).Encode(0xc0001db7e0, 0x5, 0x8, 0xc000147b30)
        external/org_golang_x_tools/go/analysis/internal/facts/facts.go:266 +0x7ad
main.checkPackage(0x147c380, 0x5, 0x5, 0x7ffeefbff88c, 0x2d, 0xc00006f4a0, 0xc00006f4d0, 0xc00006f530, 0xc00000e6d0, 0x2, ...)
        external/io_bazel_rules_go/go/tools/builders/nogo_main.go:177 +0x308
main.run(0xc00000e390, 0x36, 0x37, 0x0, 0x0)
        external/io_bazel_rules_go/go/tools/builders/nogo_main.go:80 +0x3cc
main.main()
        external/io_bazel_rules_go/go/tools/builders/nogo_main.go:54 +0x8d
INFO: Elapsed time: 1.870s, Critical Path: 0.70s
```

Reproducer is here https://github.com/atlassian/smith/compare/nogo-panic. Clone, `dep ensure`, `make update-bazel` and then `bazel test //... -- -//vendor/...`

--- a/go/tools/builders/nogo_main.go
+++ b/go/tools/builders/nogo_main.go
@@ -21,6 +21,7 @@ package main
 
 import (
 	"bytes"
+	"encoding/gob"
 	"errors"
 	"flag"
 	"fmt"
@@ -142,37 +143,49 @@ func readImportCfg(file string) (packageFile map[string]string, importMap map[st
 //
 // This implementation was adapted from that of golang.org/x/tools/go/checker/internal/checker.
 func checkPackage(analyzers []*analysis.Analyzer, packagePath string, packageFile, importMap map[string]string, stdImports map[string]bool, filenames []string) (string, []byte, error) {
+	// Register fact types and establish dependencies between analyzers.
+	actions := make(map[*analysis.Analyzer]*action)
+	var visit func(a *analysis.Analyzer) *action
+	visit = func(a *analysis.Analyzer) *action {
+		act, ok := actions[a]
+		if !ok {
+			act = &action{a: a}
+			actions[a] = act
+			for _, f := range a.FactTypes {
+				act.usesFacts = true
+				gob.Register(f)
+			}
+			act.deps = make([]*action, len(a.Requires))
+			for i, req := range a.Requires {
+				dep := visit(req)
+				if dep.usesFacts {
+					act.usesFacts = true
+				}
+				act.deps[i] = dep
+			}
+		}
+		return act
+	}
+
+	roots := make([]*action, 0, len(analyzers))
+	for _, a := range analyzers {
+		roots = append(roots, visit(a))
+	}
+
+	// Load the package, including AST, types, and facts.
 	imp := newImporter(importMap, packageFile, stdImports)
 	pkg, err := load(packagePath, imp, filenames)
 	if err != nil {
 		return "", nil, fmt.Errorf("error loading package: %v", err)
 	}
-
-	// Construct the action graph.
-	var roots []*action
-	actions := make(map[*analysis.Analyzer]*action)
-
-	var visit func(a *analysis.Analyzer) *action
-	visit = func(a *analysis.Analyzer) *action {
-		act, ok := actions[a]
-		if !ok {
-			act = &action{a: a, pkg: pkg}
-
-			// Add a dependency on each required analyzers.
-			for _, req := range a.Requires {
-				act.deps = append(act.deps, visit(req))
-			}
-			actions[a] = act
-		}
-		return act
+	for _, act := range actions {
+		act.pkg = pkg
 	}
 
-	for _, analyzer := range analyzers {
-		action := visit(analyzer)
-		roots = append(roots, action)
-	}
-
+	// Execute the analyzers.
 	execAll(roots)
+
+	// Process diagnostics and encode facts for importers of this package.
 	diagnostics := checkAnalysisResults(roots, pkg)
 	facts := pkg.facts.Encode()
 	return diagnostics, facts, nil
@@ -191,6 +204,7 @@ type action struct {
 	inputs      map[*analysis.Analyzer]interface{}
 	result      interface{}
 	diagnostics []analysis.Diagnostic
+	usesFacts   bool
 	err         error
 }
 
@@ -200,13 +214,12 @@ func (act *action) String() string {
 
 func execAll(actions []*action) {
 	var wg sync.WaitGroup
+	wg.Add(len(actions))
 	for _, act := range actions {
-		wg.Add(1)
-		work := func(act *action) {
+		go func(act *action) {
 			act.exec()
 			wg.Done()
-		}
-		go work(act)
+		}(act)
 	}
 	wg.Wait()
 }

--- a/tests/core/nogo/vet/BUILD.bazel
+++ b/tests/core/nogo/vet/BUILD.bazel
@@ -63,6 +63,7 @@ go_library(
     name = "has_errors",
     srcs = ["has_errors.go"],
     importpath = "haserrors",
+    deps = [":fmtwrap"],
 )
 
 go_library(
@@ -70,4 +71,10 @@ go_library(
     srcs = ["no_errors.go"],
     cgo = True,
     importpath = "noerrors",
+)
+
+go_library(
+    name = "fmtwrap",
+    srcs = ["fmtwrap.go"],
+    importpath = "fmtwrap",
 )

--- a/tests/core/nogo/vet/fmtwrap.go
+++ b/tests/core/nogo/vet/fmtwrap.go
@@ -1,0 +1,7 @@
+package fmtwrap
+
+import "fmt"
+
+func Printf(format string, args ...interface{}) {
+	fmt.Printf(format, args...)
+}

--- a/tests/core/nogo/vet/has_errors.go
+++ b/tests/core/nogo/vet/has_errors.go
@@ -3,7 +3,7 @@ package haserrors
 // +build build_tags_error
 
 import (
-	"fmt"
+	"fmtwrap"
 	"sync/atomic"
 )
 
@@ -15,6 +15,6 @@ func Foo() bool {
 	if F == nil { // nilfunc error.
 		return false
 	}
-	fmt.Printf("%b", "hi") // printf error.
-	return true || true    // redundant boolean error.
+	fmtwrap.Printf("%b", "hi") // printf error.
+	return true || true        // redundant boolean error.
 }

